### PR TITLE
`ultralytics 8.0.235` YOLOv8 OBB train, val, predict and export

### LINKED ISSUE #4242: Help Wanted 🙏 Add Oriented Bounding Boxes (OBB) Task with DOTA v2 Training
Hello YOLOv8 Enthusiasts,

We are reaching out to the talented community to help add support for [DOTA v2](https://captain-whu.github.io/DOTA/dataset.html) training with Oriented Bounding Boxes (OBB) for our YOLOv8 repository. This addition will notably enhance our detection capabilities for remote sensing imagery, enabling superior performance for real-world applications.

Related to issue https://github.com/ultralytics/ultralytics/issues/624 and PR https://github.com/ultralytics/ultralytics/pull/1977

## Project Description:

Our primary objective with this issue is to integrate the DOTA v2 dataset into our YOLOv8 training pipeline, with a focus on Oriented Bounding Boxes. This demands crafting methods or functions to accommodate the unique data format of DOTA v2, and seamlessly incorporate it into our existing training framework.

## Example code:

```python
from yolov8 import YOLO

# Initialize a model
model = YOLO('yolov8n-obb.pt')  # load a pretrained OBB model

# Train the model on DOTA v2 with OBB
model.train(data='DOTAv2.yaml', epochs=30, imgsz=640)
```

In this envisaged example, the `data` argument would accept a configuration file specifically crafted for the DOTA v2 dataset with OBB, `'DOTAv2.yaml'`. This configuration file should encompass all essential details to fetch, process, and utilize the dataset during training.

## Expected Deliverables:

- Code that processes and manages the DOTA v2 dataset with Oriented Bounding Boxes.
- Comprehensive test cases to validate the flawless functioning of the added feature.
- Amended documentation, enlightening users on the ins and outs of this fresh capability.

## How to contribute:

1. Express your interest in this endeavor by commenting on this issue.
2. Fork the YOLOv8 repository.
3. Generate a fresh branch in your forked repository.
4. Execute the necessary code modifications.
5. Develop tests that corroborate the effectiveness of your code.
6. Revise the README and affiliated documentation accordingly.
7. Submit a detailed pull request, illuminating your changes, their relevance, and how they enhance the repository.

Every contribution, regardless of its magnitude, is deeply appreciated. Your efforts contribute massively to the broader community, and every input is treated with respect.

## Community Support:

To amplify its visibility, this issue will be pinned. If you need support or have inquiries, do not hesitate to contact us either on this thread or via our [Discord](https://ultralytics.com/discord).

Gratitude and best wishes on your coding journey!

Warm Regards,
The YOLOv8 Team

### PR DESCRIPTION
**Pull request by glenn-jocher**
Supersedes original OBB PR #1977

IMPORTANT: THIS PR IS STILL UNDER CONSTRUCTION AND OBB TRAINING DOES **NOT** WORK CORRECTLY YET.

### Summary
🔄📦📐

The pull request adds support for the oriented bounding box (obb) task, which is a new detection model that can handle rotated objects. It introduces new classes, functions, and arguments to implement the obb model and its data processing, training, validation, and prediction functionalities. It also refactors and simplifies some existing code for handling different input image types and formats. The main files affected are `ultralytics/models/yolo/model.py`, `ultralytics/models/yolo/obb/*`, `ultralytics/data/*`, and `ultralytics/cfg/*`.

> _The pull request adds a new feature_
> _To detect objects with a skew or a skewer_
> _It uses `obb` labels and models_
> _And modifies some `postprocess` modules_
> _To handle different image types and scales with fewer_

### Walkthrough
*  Add support for oriented bounding box (obb) task, which allows the user to train and predict using obb model and data ([link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-c02fb3f9e86e97bf85f07211aed56aa04774504a61eeff07a08471bbdea14556L18-R36), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-7e6b10c61fd65eca79a0acc61eeff14a35dc5654325f44c687251c61393d2ae0R1-R46), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L16-R16), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528R372-R402), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L444-R485), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L454-R493), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528R740), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528R748), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L737-R780), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-e8bb4966653e657a3bfc4a4c892ccdcf3c486a982016d2e51631b5b9393c2c5fR88), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L34-R39), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L59-R63), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L73-R75), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L103-R106), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90R162), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L176-R190), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911cL25-R25), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911cL62-R66), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911cL85-R96), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911cR104-R108), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-ec08ff3087043c4949ebd4c16ec17073bdb827e25ff1840d1eb7c87841b8911cL126-R146), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955L92-R92), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R100), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955R608-R669), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL42-R46), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-d85395b639979a65be5ffda459dd071688c28222abea583cf25df1780d0e0d07L4-R5), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-d85395b639979a65be5ffda459dd071688c28222abea583cf25df1780d0e0d07L36-R41), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-9803362fccd01b7685da3755bfe2fcaee2fa4ab721dfa49ab54f32fe18be2519R1-R7), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-6652ecbc1d9c6d692a086d27b9e184e1d81dd4d4471d09feab7dcedad5c2d486R1-R28), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-b0d02d6e1e01adda144f1959393dd46500081f41c560ef947d4b203e28b37122R1-R42), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-dc796eab82f77289ce5740f0b061ac0a734e5914d1245bc51248ebcf7674b39eR1-R27))
*  Simplify the logic for handling different types of input images in various predictor classes, using a variable `is_list` to indicate whether the images are a list or a torch tensor, and avoiding repeating the `path` variable assignment ([link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-7c48ede77b870dae7664b49a45c6f9aa0c835fd37fa9e084e1e5bb05b4d27ed8L35-R49), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-cb0a8d6008c9163906be795736ce0434301c4228f770740e3fb91c9177c840c1L27-R33), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-752d7deb4a6279bd12373b285f7be5c587ca4805bd62976d0e0c577d6b0acf3aR31), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-752d7deb4a6279bd12373b285f7be5c587ca4805bd62976d0e0c577d6b0acf3aL38-R45), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-abae5c0cfe4087dc193f98a5d67d7c726ac769ae8a58b3abda232a7657c69edeL321-R323), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-abae5c0cfe4087dc193f98a5d67d7c726ac769ae8a58b3abda232a7657c69edeL330-R332), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-9dded00f5e51a32b00c6266aa445990bae22fe6ea0bbb9d2c8898f4fcdb8cb23L35-R39), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-383afc92a2a66556c136cd357f28db596eed25b3472abe26b7e778208a365571L41-R49))
*  Change the assertion and warning messages in the `Boxes` and `Masks` classes to be more clear and concise ([link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955L410-R411), [link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-24af401e53ffc5ce49197395e1e3d7111344ad31b08cbb2bee9377614da98955L493-R494))
*  Remove an unnecessary import of the `torch` module in the `ultralytics/models/yolo/detect/predict.py` file ([link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-7bb0bb8a018eb6885d35c69879091dc25b89a04046f5b3ce71e82ddb1c6f65b7L3-R7))
*  Remove a redundant comment in the `FastSAMPredictor` class that did not reflect the actual code functionality ([link](https://github.com/ultralytics/ultralytics/pull/4499/files?diff=unified&w=0#diff-7c48ede77b870dae7664b49a45c6f9aa0c835fd37fa9e084e1e5bb05b4d27ed8L18))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
New OBB (Oriented Bounding Box) detection models introduced for improved aerial imagery processing.

### 📊 Key Changes
- Addition of OBB detection models with DOTA dataset support. 
- Inclusion of pre-trained model weights and new documentations pages.
- Enhanced model configuration for OBB tasks.

### 🎯 Purpose & Impact
The new features are intended to enhance the detection of oriented objects in aerial images, benefiting applications like satellite image analysis and surveillance. The added models bring improvements to Ultralytics offerings, potentially attracting new users and use-cases, particularly those dealing with geospatial data.

### CODE REVIEW

**Comment by WangQvQ** [comment]
Looking forward to it very much

**Comment by glenn-jocher** [comment]
@WangQvQ thank you for your interest in YOLOv8! We're also excited about the development work on this project. Stay tuned for updates and future enhancements. Your enthusiasm and support are greatly appreciated.

**Comment by jaydeboer** [comment]
+1 from me as well.  I am checking in almost daily to see the progress.  Thanks for all the effort!

**Comment by GAcherQuaze** [comment]
Quite excited seeing that PR !
I just saw that theta_based bboxes are being removed in favour of segments, which is understandable as it keeps things lean. However, it feels like it might pose a limitation : for users interested in inferring an angle/heading in a 0-360 range, segment-based annotations can lead to ambiguities. 
In the current code, the neural network uses xywha format, so the angular ambiguity can probably be mitigated by thinking of a way to standardize the ordering of segments in the annotation (for instance, theta could be the angle between a vertical upwards vector ; and the vector going from the barycenter of vertices to the middle of the *First* segment). 
Thanks for the great start, and please let me know if/how I can help

**Comment by glenn-jocher** [comment]
@jaydeboer hey there,

Thank you for your interest in YOLOv8! We're happy to hear that the PR has struck a chord with you!

You've raised a good point about the shift from theta-based bounding boxes to segments. Rest assured, it's definitely something we've considered, and we agree that for certain use cases, such as estimating an object's orientation in a 0-360 range, the ambiguity could become an issue.

Currently, we could possibly manage these ambiguities by a method similar to what you suggested, a standardized ordering of segments during annotation.

We're always open to feedback and ideas that could help improve the codebase, so we appreciate your contribution. Do feel free to put forward any suggestions or make contributions if you'd like! Your input is highly valued.

Thank you for engaging with us, and stay tuned for more updates!

**Comment by karthikyerram** [comment]
@glenn-jocher Is the obb feature ready to use? If yes can you point me to the right place?

**Comment by Laughing-q** [comment]
@karthikyerram it's still working in progress

**Comment by karthikyerram** [comment]
@Laughing-q can I know what is the expected timeline for this?
Thank you. Looking forward to your answer.

**Comment by Laughing-q** [comment]
@karthikyerram there's no exact timeline currently, but it's almost done, just needs some experiments. :)

**Comment by karthikyerram** [comment]
@Laughing-q  Our tool would be highly dependent on this feature. So to plan things well, could you please give a safe range? This could be really helpful for us to understand our timeline. Thank you

**Comment by JunYanOOO** [comment]
Hello, I would like to inquire if the development of the OBB feature has been completed?
Is the usage based on the training method of DOTA v2?
Much appreciated.

**Comment by glenn-jocher** [comment]
@JunYanOOO hello! The OBB feature is still under active development and is not yet completed. It is being tailored to handle datasets like DOTA v2 effectively. Stay tuned for updates, and thank you for your interest!

**Comment by glenn-jocher** [comment]
@Laughing-q @AyushExel @fcakyon ok merging this PR as publishing as 8.0.235! If there are any remaining issues we can address in follow-on fix PRs :)
